File: src/main/java/lechuck/intellij/vars/VariablesTextFieldWithBrowseButton.java
package lechuck.intellij.vars;

import com.google.common.collect.ImmutableMap;
import com.intellij.icons.AllIcons;
import com.intellij.openapi.keymap.KeymapUtil;
import com.intellij.openapi.ui.TextFieldWithBrowseButton;
import com.intellij.openapi.util.NlsContexts;
import com.intellij.openapi.util.text.StringUtil;
import com.intellij.ui.DocumentAdapter;
import com.intellij.ui.UserActivityProviderComponent;
import com.intellij.util.containers.ContainerUtil;
import org.jetbrains.annotations.NotNull;

import javax.swing.*;
import javax.swing.event.ChangeEvent;
import javax.swing.event.ChangeListener;
import javax.swing.event.DocumentEvent;
import java.awt.event.InputEvent;
import java.awt.event.KeyEvent;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

public class VariablesTextFieldWithBrowseButton extends TextFieldWithBrowseButton implements UserActivityProviderComponent {

    protected VariablesData myData = VariablesData.DEFAULT;
    private final List<ChangeListener> myListeners = ContainerUtil.createLockFreeCopyOnWriteList();

    public VariablesTextFieldWithBrowseButton() {
        super();
        addActionListener(e -> {
            setVars(VariablesTable.parseVarsFromText(getText()));
            createDialog().show();
        });
        getTextField().getDocument().addDocumentListener(new DocumentAdapter() {
            @Override
            protected void textChanged(@NotNull DocumentEvent e) {
                if (!StringUtil.equals(stringifyVars(myData), getText())) {
                    Map<String, String> textVars = VariablesTable.parseVarsFromText(getText());
                    myData = myData.with(textVars);
                    fireStateChanged();
                }
            }
        });
    }

    @NotNull
    protected VariablesDialog createDialog() {
        return new VariablesDialog(this);
    }

    /**
     * @return unmodifiable Map instance
     */
    @NotNull
    public Map<String, String> getVars() {
        return myData.getVars();
    }

    /**
     * @param vars Map instance containing variables
     *             (iteration order should be reliable user-specified, like {@link LinkedHashMap} or {@link ImmutableMap})
     */
    public void setVars(@NotNull Map<String, String> vars) {
        setData(myData.with(vars));
    }

    @NotNull
    public VariablesData getData() {
        return myData;
    }

    public void setData(@NotNull VariablesData data) {
        VariablesData oldData = myData;
        myData = data;
        setText(stringifyVars(data));
        if (!oldData.equals(data)) {
            fireStateChanged();
        }
    }

    @NotNull
    @Override
    protected Icon getDefaultIcon() {
        return AllIcons.General.InlineVariables;
    }

    @NotNull
    @Override
    protected Icon getHoveredIcon() {
        return AllIcons.General.InlineVariablesHover;
    }

    @NotNull
    protected String stringifyVars(@NotNull VariablesData varData) {
        if (varData.getVars().isEmpty()) {
            return "";
        }
        StringBuilder buf = new StringBuilder();
        for (Map.Entry<String, String> entry : varData.getVars().entrySet()) {
            if (!buf.isEmpty()) {
                buf.append(";");
            }
            buf.append(StringUtil.escapeChar(entry.getKey(), ';'))
                    .append("=")
                    .append(StringUtil.escapeChar(entry.getValue(), ';'));
        }
        return buf.toString();
    }

    @Override
    public void addChangeListener(@NotNull ChangeListener changeListener) {
        myListeners.add(changeListener);
    }

    @Override
    public void removeChangeListener(@NotNull ChangeListener changeListener) {
        myListeners.remove(changeListener);
    }

    private void fireStateChanged() {
        for (ChangeListener listener : myListeners) {
            listener.stateChanged(new ChangeEvent(this));
        }
    }

    protected static List<Variable> convertToVariables(Map<String, String> map) {
        return ContainerUtil.map(map.entrySet(),
                entry -> new Variable(entry.getKey(), entry.getValue()));
    }

    @Override
    protected @NotNull @NlsContexts.Tooltip String getIconTooltip() {
        return "Edit variables (" +
                KeymapUtil.getKeystrokeText(KeyStroke.getKeyStroke(KeyEvent.VK_ENTER, InputEvent.SHIFT_DOWN_MASK)) + ")";
    }
}

File: src/main/kotlin/lechuck/intellij/vars/VariablesTextFieldWithBrowseButton.kt
package lechuck.intellij.vars

import com.intellij.icons.AllIcons
import com.intellij.openapi.keymap.KeymapUtil
import com.intellij.openapi.ui.TextFieldWithBrowseButton
import com.intellij.openapi.util.NlsContexts
import com.intellij.openapi.util.text.StringUtil
import com.intellij.ui.DocumentAdapter
import com.intellij.ui.UserActivityProviderComponent
import com.intellij.util.containers.ContainerUtil
import org.jetbrains.annotations.NotNull
import java.awt.event.InputEvent
import java.awt.event.KeyEvent
import javax.swing.Icon
import javax.swing.KeyStroke
import javax.swing.event.ChangeEvent
import javax.swing.event.ChangeListener
import javax.swing.event.DocumentEvent

class VariablesTextFieldWithBrowseButton : TextFieldWithBrowseButton(), UserActivityProviderComponent {
    private var myData: VariablesData = VariablesData.DEFAULT
    private val myListeners = ContainerUtil.createLockFreeCopyOnWriteList<ChangeListener>()

    init {
        addActionListener {
            setVars(VariablesTable.parseVarsFromText(text))
            createDialog().show()
        }
        textField.document.addDocumentListener(object : DocumentAdapter() {
            override fun textChanged(@NotNull e: DocumentEvent) {
                if (!StringUtil.equals(stringifyVars(myData), text)) {
                    val textVars: Map<String, String> = VariablesTable.parseVarsFromText(text)
                    myData = myData.with(textVars)
                    fireStateChanged()
                }
            }
        })
    }

    private fun createDialog(): VariablesDialog {
        return VariablesDialog(this)
    }

    /**
     * @return unmodifiable Map instance
     */
    fun getVars(): Map<String, String> {
        return myData.vars
    }

    /**
     * @param vars Map instance containing variables
     *             (iteration order should be reliable user-specified, like [LinkedHashMap] or [ImmutableMap])
     */
    fun setVars(vars: Map<String, String>) {
        setData(myData.with(vars))
    }

    fun getData(): VariablesData {
        return myData
    }

    fun setData(data: VariablesData) {
        val oldData = myData
        myData = data
        text = stringifyVars(data)
        if (oldData != data) {
            fireStateChanged()
        }
    }

    override fun getDefaultIcon(): Icon {
        return AllIcons.General.InlineVariables
    }

    override fun getHoveredIcon(): Icon {
        return AllIcons.General.InlineVariablesHover
    }

    private fun stringifyVars(varData: VariablesData): String {
        if (varData.vars.isEmpty()) {
            return ""
        }
        val buf = StringBuilder()
        for ((key, value) in varData.vars) {
            if (buf.isNotEmpty()) {
                buf.append(";")
            }
            buf.append(StringUtil.escapeChar(key, ';'))
                .append("=")
                .append(StringUtil.escapeChar(value, ';'))
        }
        return buf.toString()
    }

    override fun addChangeListener(l: ChangeListener) {
        myListeners.add(l)
    }

    override fun removeChangeListener(l: ChangeListener) {
        myListeners.remove(l)
    }

    private fun fireStateChanged() {
        for (listener in myListeners) {
            listener.stateChanged(ChangeEvent(this))
        }
    }


    override fun getIconTooltip(): @NlsContexts.Tooltip String {
        val shortcut = KeymapUtil.getKeystrokeText(
            KeyStroke.getKeyStroke(KeyEvent.VK_ENTER, InputEvent.SHIFT_DOWN_MASK)
        )
        return "Edit variables ($shortcut)"
    }

    companion object {
        fun convertToVariables(map: Map<String, String>): List<Variable> {
            return ContainerUtil.map(map.entries) { (key, value) -> Variable(key, value) }
        }
    }
}